### PR TITLE
Fix suppressed policy errors

### LINKED ISSUE #18566: PolicyError are replaced by ForbiddenError
### Required System information

- Node.js version: v16
- NPM version:
- Strapi version: v4.14.3
- Database: PostgreSQL
- Operating system: Linux
- Is your project Javascript or Typescript: Typescript
 
### Describe the bug

It seems that when throw a policyError, the response are always a ForbiddenError without PolicyError custom message.

### Steps to reproduce the behavior
Create and use any policy.
Throw a new PolicyExceptionError
Call the endpoint that uses that policy.
You'll receive a Forbidden Exception

This example from documentation

https://docs.strapi.io/dev-docs/backend-customization/examples/policies#sending-custom-errors-through-policies

### Expected behavior
I expect receive PolicyError,

[@strapi+strapi+4.14.3.patch](https://github.com/strapi/strapi/files/13166861/%40strapi%2Bstrapi%2B4.14.3.patch)

### PR DESCRIPTION
**Pull request by innerdvations**
### What does it do?

Allows UnauthorizedError and ForbiddenError in the auth middleware to bubble up rather than catching them and sending our own empty error

TODO:

- [ ] write test that Policy throwing a PolicyError should bubble up

### Why is it needed?

PolicyError thrown in policies is currently being suppressed and returning a generic ForbiddenError instead

### How to test it?

- create a policy
- have it throw a PolicyError

- without this PR: see that the api currently returns a ForbiddenError instead
- using this PR: it returns the correct PolicyError

Tests are in progress

### Related issue(s)/PR(s)

Fixes #18566

### CODE REVIEW

**Review by Convly — COMMENTED** [review]
I'm not sure anymore as I can't find the definition of `ctx.unauthorized` or forbidden in the codebase anymore. But I believe this was made on purpose to avoid leaking information from the internal Unauthorized or Forbidden error classes.

I'll let @alexandrebodin have the last word as he probably remembers better than me this part.

**Comment by innerdvations** [comment]
If we just want to fix this specific case to be safe, my initial version was actually just a `&& !(error typeof errors.PolicyError)` added to the forbiddenerror clause, which worked fine to pinpoint the referenced bug but felt a bit hacky. I'll revert to that for now to minimize impact, and if there are any other edge cases remaining we can handle them later.

**Comment by Convly** [comment]
> I'm not sure anymore as I can't find the definition of `ctx.unauthorized` or forbidden in the codebase anymore. But I believe this was made on purpose to avoid leaking information from the internal Unauthorized or Forbidden error classes.

To clarify, I'm open to allowing PolicyErrors to throw their custom errors, but we must be careful with what it implies for existing policies that might not want to have their exact content returned to the user (e.g. "user already exists", etc...)

**Comment by alexandrebodin** [comment]
Hum I would change the documentation instead of changing the behavior 🤔 
I don't think we should start leaking any policy error info that we didn't before without considering this at least a breaking change and at most a security risk. 

If you throw a custom ApplicationError we will return the error you want so I would probably document this instead 🤔 and keep the policy error become a forbidden error with no leaked information.

**Comment by innerdvations** [comment]
> Hum I would change the documentation instead of changing the behavior 🤔 I don't think we should start leaking any policy error info that we didn't before without considering this at least a breaking change and at most a security risk.
> 
> If you throw a custom ApplicationError we will return the error you want so I would probably document this instead 🤔 and keep the policy error become a forbidden error with no leaked information.

Tthe breaking change came at some point during v4, we've documented that PolicyErrors are visible to the content API as the method to publicly expose error messages and that `return false` should be used from a policy if you want to fail without providing additional information -- there are blog posts, forum posts, etc, explaining this method, which seems to have previously worked, it's not just one doc. For example, here's a post showing that it worked that way from March 2022: https://forum.strapi.io/t/strapi4-policies-can-find-body-in-policycontext-throw-error-in-policy/17014/3

If you're worried users have used it counter to documentation to throw private data, we can instead update the documentation to show to use ApplicationError. But in that case, we should try to improve that in v5 if we have time and allow any error to be thrown (ie, any http status and message) and explicitly make it public.

**Comment by innerdvations** [comment]
@alexandrebodin 

Well, this has quite an interesting history.

ForbiddenErrors and AuthorizedErrors have always (since 2021) been converted to generic while allowing ApplicationErrors to pass through.

PolicyError was introduced in [this PR ](https://github.com/strapi/strapi/pull/12080) with two explicit purposes:
- fix the status code of returning 'false' from a policy
- provide a way to return a message with a forbidden status to the content api from policies

It was originally implemented with PolicyError extending ApplicationError (and would have worked that way during the initial testing), then before merging it was changed to ForbiddenError in [this commit](https://github.com/strapi/strapi/commit/4dbb64e8a2d08f5281434daa18cff70141fe6c8b)

But... that means the feature apparently never worked as intended (and you can see someone complaining about that on the merged PR) despite being documented.

What do you think about:
- in v5, allow PolicyError to pierce through and move the existing documentation for v4 to v5 docs
- replace the v4 docs with "If you need a public message, throw an ApplicationError" to avoid leaking anything from a user who implemented something while it was broken

File: packages/core/strapi/src/services/server/compose-endpoint.ts [hunk 1]
```diff
@@ -36,17 +37,33 @@ const createAuthorizeMiddleware =
...
     } catch (error) {
+      /**
+       * Allow policy errors (thrown by user policies) to bubble up, since in those cases we are
+       * not leaking data, but rather enforcing a policy.
+       *
+       * This must come before ForbiddenError check because they are a subclass of ForbiddenError
+       *
+       * TODO: this could be improved by adding a PrivatePolicyError or a `private` property to PolicyError,
+       * to indicate that it should not be returned to the client, but currently PolicyErrors are expected
+       * to return only publicly safe error messages.
```
Comment (innerdvations) [inline packages/core/strapi/src/services/server/compose-endpoint.ts:50]:
```suggestion
```

**Comment by alexandrebodin** [comment]
Hey considering this has probably always been broken I would go with option 1. We are not investing in fixing v4 so much as we are making the necessary breaking changes in v5 right now so better document a good way to return public errors for people wanting it than risking leaking things even if it's documented otherwise. Not worth the risk so late in the v4 lifetime.

**Comment by innerdvations** [comment]
Closing to target v5 with a similar change

**Pull request closed without merging**